mark.c

```diff
@@ -841,18 +841,18 @@ GC_INNER mse * GC_mark_from(mse *mark_stack_top, mse *mark_stack,
         /* marking for that chunk.                                      */
         /* If the pointer is out of bounds at the upper-end, reset      */
         /* the pointer within bounds and continue from there            */
+        if (cheri_getaddress(limit) + sizeof(ptr_t) - 1
+              >= (cheri_getbase(limit) + cheri_getlength(limit))) {
+          /* reset pointer to within boundary of current_p */
+          /* from which limit is derived                   */
+          limit = cheri_setaddress( current_p, ((cheri_getbase(limit)
+                                    + cheri_getlength(limit) - sizeof(ptr_t))
```

Comment:
What happens if the length of `limit` is less than `sizeof(ptr_t)` -- will this cause us to get a capability with an address lower than the permitted bounds?

Reply:
I do not think this is an issue as far as I can see. 

1. In this case, limit is derived from current_p at the outset so `base(current_p) = base(limit)`
2. As far as I understand ,
  - For large objects, we will have at least 128 objects. (i.e. 2048 bytes) 
  - For small objects this should be at least  1   `granule` i.e. greater than 16 bytes. 
3. Even if  length(limit)  and results in a `address(limit) < base(current_p)` (because *limit* is derived from *current_p*) , while this will be set to that address, upon first encountering the first guard within the `for(;;)` loop, this will exit. 

However, you have prompted another issue, i.e. will the `cheri_setaddress()` raise a SIGPROT exception or allow execution to continue until the pointer is derefernced. I have checked on the `riscv` port and it allows execution to continue until the pointer is dereferenced.  But that might not always be portable. Do you know otherwise @ltratt, @jacobbramley and/or @jeremysinger ?

Reply:
> Even if length(limit) and results in a address(limit) < base(current_p) (because limit is derived from current_p) , while this will be set to that address, upon first encountering the first guard within the for(;;) loop, this will exit.

OK, I buy this is a back-up. Does that mean the first thing we should do in each iteration is check whether we have a valid capability or not?

Reply:
yes, the code currently checks whether the tag is set or not and whether we have undershot the base of the capability , before dereferencing it.

Reply:
OK, so I'm now happy. Let's wait for @jacobbramley to comment, but hopefully we're nearly there!

Reply:
I'm not sure about the portability of the CHERI API, but on Morello at least, `cheri_setaddress()` won't abort. There are three cases:

1. The result is within bounds. This is always encodable (since we aren't changing the bounds), and not of concern.
2. The result is out of bounds, but encodable. This is important for supporting both standard and de-facto C usage, and is extensively described in CHERI papers (especially the CHERI-Concentrate one, if I remember correctly). This behaviour can probably be considered as portable as anything else in a research architecture.
3. The result is out of bounds, and not encodable. Morello clears the tag on the result, and I think other CHERI implementations do the same. If there's any doubt about portability, this is where it will be, but this arrangement is convenient in many respects (e.g. for use cases like this one), and I would expect plenty of software to rely on it.

On that note, is there any guarantee that `current_p` actually points inside its own bounds? That was part of my motivation behind suggesting that you constrain the search to the bounds (so you only search the intersection of Boehm GC's idea of the object and the capability bounds), since then you don't have to think about complicated bounds situations.

Reply:
> On that note, is there any guarantee that current_p actually points inside its own bounds?

I *think* that @djichthys is checking the CHERI tag validity so this is implicitly checked?

Reply:
Consider a `current_p` that with an address higher than its limit.
* We hit this code because `limit` will inherit that property.
* The `cheri_setaddress()` call forces `address(limit)` to point to the last `ptr_t` in its bounds. _However_, this could be below the address of `current_p`.
* We then fail the assertion on entering the `for(;;)` loop below.
I've quite possibly confused something, and might be looking at some intermediate patch rather than the end result.